### PR TITLE
[PT-D][BE][TP perf 1/N] Get rid of unnecessary collectives in Embedding/EmbeddingBag and use autograd-enabled collectives

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81853

These two ops (Embedding and EmbeddingBag for ShardedTensor) especially for row-wise sharding is very inefficient and hard to fit in the concept of future design. So this PR is trying to:
1. Remove all unnecessary collective communications. Only one gather and one reduce(or reduce scatter) is needed.
2. Use auto-grad enabled collectives so that we can use these ops in real model training.
3. Some minor code cleaning
4. Treat input differently when it's replicated tensor. (Will add more for this for the next few PRs).

Differential Revision: [D37965687](https://our.internmc.facebook.com/intern/diff/D37965687/)

### CODE REVIEW

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py [hunk 1]
```diff
@@ -226,170 +225,13 @@ def _result_distribute_with_col_rearrange(
     return output.index_select(0, torch.tensor(indices, device=output.device))
 
 
-def _handle_row_wise_lookup_distribute(
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:229]: does it mean that rowwise look up is not needed anymore?

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py [hunk 2]
```diff
@@ -39,34 +41,25 @@ def sharded_embedding(types, args, kwargs, pg):
     4 GPUs creating 3 shard of (3 x 17) and 1 shard of (1 x 17).
     The algorithm is as follows:
 
-    1. First the input is flattened to 1D and gets sorted so that we can distribute
-       them to the corresponding rank. For example if the given input is
+    1. First the input is broadcasted to all ranks, since this is SPMD we
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py:44]: should this be an all_gather or broadcast, looks confusing to me, usually if input is sharded, we all_gather, if input is not, we broadcast it to get the same values, which one we are taking here?

**Review by wanchaol — COMMENTED** [review]
Thanks for refactoring this! I did an initial pass, it looks great, have a few comments on when to use all_gather, let me know if those make sense.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:229]: Yeah this is old logic. We don't need this anymore so that we can get rid of unnecessary collective communications.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py [hunk 3]
```diff
@@ -39,34 +41,25 @@ def sharded_embedding(types, args, kwargs, pg):
...
-       all2all to send the result back to each rank.
-    6. We use the recorded indices to recover the sorted positions and reshape the
-       matrix to (4 x 6 x 17), which is what we need.
+       on rank 0.
+       Then on every rank, we will have this tensor.
+    2. Next, we mask the ID which are not stored on that rank.
+       For example on rank 0, we store ID [0, 1, 2]. We only keep the ID
+       inside the set of numbers. The rest of them will be masked to an extra row.
+       The masked matrix will be used for embedding look up and is like:
+       tensor([[4, 4, 2, 4, 4, 4],
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py:57]: 4 is the masked ID and the reason is that for the weight being sharded there is only 0, 1 and 2. We will add one more dummy entry (4); otherwise, when max_norm is specified, the result is not correct because max_norm will norm the row which is looked up.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py [hunk 4]
```diff
@@ -443,4 +283,71 @@ def _handle_max_norm_col_wise(
...
+    if (
+        padding_idx is not None
+        and padding_idx >= start_pos
+        and padding_idx < (start_pos + chunk_size)
+    ):
+        padding_idx = padding_idx - start_pos
+
+    # When max_norm is set, it will only re-norm the row being looked up.
+    padding_row = torch.zeros(
+        1, weight.size(1), device=gather_inp.device, dtype=weight.dtype
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:351]: No, this is the extra row we padded to the weight and weight is a sharded tensor it does not have device implemented.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py [hunk 5]
```diff
@@ -215,11 +206,11 @@ def _handle_col_wise_sharding(
 
     Returns: final result of lookup.
     """
-    gathered_inputs = None
+    gathered_inputs = all_gather(input, group=pg)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py:209]: Oh, sure. Let me add that.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding_bag.py [hunk 6]
```diff
@@ -377,443 +352,132 @@ def _handle_row_wise_sharding(
...
+        if input.dim() > 1 and per_sample_weights is None:
+            # allgather the inputs first for non Replicated Tensor.
+            gather_inp = _all_gather_base_input(input, pg)
+        else:
+            (
+                gathered_inputs,
+                gathered_per_sample_weights,
+                gathered_offsets,
+            ) = _all_gather_embedding_bag_input(input, per_sample_weights, offsets, pg)
+            cat_dim = 0 if input.dim() > 1 else -1
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding_bag.py:365]: This is for differentiating the case when dim == 1 and dim > 1. Dim will not < 1.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py [hunk 7]
```diff
@@ -406,24 +248,22 @@ def _handle_max_norm_col_wise(
...
         local_shard_norm_renormed: local_shard re-normed to max_norm if the norm is larger
             than it.
-        gathered_inputs: list of inputs from all ranks.
+
     """
     norm_type = norm_type if norm_type is not None else 2.0
-    # allgather the inputs first.
-    gathered_inputs = [torch.zeros_like(input) for _ in range(world_size)]
-    dist.all_gather(gathered_inputs, input, group=pg)
     unique_inp = torch.unique(torch.cat(gathered_inputs))
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:260]: Let me think about this..

[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:260]: After a second thought, we probably need to keep using all_gather for col-wise sharding because we need to make it a list and do all the rearrangement. However, when we migrating to DT, this should not be a problem at all.

**Review by wanchaol — APPROVED** [review]
looks good! please fix the lint and make sure multigpu jobs pass before landing (add a ciflow/periodic to trigger it)
